engine/access/ingestion/tx_error_messages/requester.go

```diff
@@ -0,0 +1,145 @@
+package tx_error_messages
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	execproto "github.com/onflow/flow/protobuf/go/flow/execution"
+	"github.com/rs/zerolog"
+	"github.com/sethvargo/go-retry"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-go/engine/common/rpc/convert"
+
+	"github.com/onflow/flow-go/engine/access/rpc/backend"
+	"github.com/onflow/flow-go/engine/common/rpc"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// RequesterConfig contains the retry settings for the tx error messages fetch.
+type RequesterConfig struct {
+	// the initial delay used in the exponential backoff for failed tx error messages download
+	// retries.
+	RetryDelay time.Duration
+	// the max delay used in the exponential backoff for failed tx error messages download.
+	MaxRetryDelay time.Duration
+}
+
+type Requester struct {
+	logger                     zerolog.Logger
+	config                     *RequesterConfig
+	backend                    *backend.Backend
+	execNodeIdentitiesProvider *rpc.ExecutionNodeIdentitiesProvider
+	executionResult            *flow.ExecutionResult
+}
+
+func NewRequester(
+	logger zerolog.Logger,
+	config *RequesterConfig,
+	backend *backend.Backend,
+	execNodeIdentitiesProvider *rpc.ExecutionNodeIdentitiesProvider,
+	executionResult *flow.ExecutionResult,
+) *Requester {
+	return &Requester{
+		logger:                     logger,
+		config:                     config,
+		backend:                    backend,
+		execNodeIdentitiesProvider: execNodeIdentitiesProvider,
+		executionResult:            executionResult,
+	}
+}
+
+// Request fetches transaction error messages for the specific
+// execution result this requester was configured with.
+func (r *Requester) Request(ctx context.Context) ([]flow.TransactionResultErrorMessage, error) {
+	backoff := retry.NewExponential(r.config.RetryDelay)
+	backoff = retry.WithCappedDuration(r.config.MaxRetryDelay, backoff)
+	backoff = retry.WithJitterPercent(15, backoff)
+
+	blockID := r.executionResult.BlockID
+	resultID := r.executionResult.ID()
+
+	var (
+		errMessages []flow.TransactionResultErrorMessage
+		lastErr     error
+	)
+
+	attempt := 0
+	err := retry.Do(ctx, backoff, func(context.Context) error {
+		if attempt > 0 {
+			r.logger.Debug().
+				Str("block_id", blockID.String()).
+				Str("result_id", resultID.String()).
+				Uint64("attempt", uint64(attempt)).
+				Msgf("retrying download")
+		}
+		attempt++
+
+		var err error
+		errMessages, err = r.request(ctx, blockID, resultID)
+		if errors.Is(err, rpc.ErrNoENsFoundForExecutionResult) || status.Code(err) != codes.Canceled {
+			lastErr = err
+			return retry.RetryableError(err)
+		}
+
+		lastErr = err
+		return err
+	})
+
+	if err != nil {
+		return nil, lastErr
```

Comment:
doesn't `retry.Do` return the last error? can we just return `err` here instead of tracking another variable